### ISSUE TITLE
Refactor architecture to use pydantic

### ISSUE THREAD

**Issue by bvreede** [elpaco-escience/scikit-talk#45]
There are a number of limitations to the (data) classes we are using in the current architecture, and especially since we need to flexibly transition between json and our class structure, and want to be flexible to adapt to new analytical methods, I am strongly leaning to applying Pydantic instead of the current architecture.

@carschno can we take some time to explore/implement this together?

**Comment by carschno** [comment]
I agree that PyDantic would be a better solution for serialization and deserialization, especially when it comes to validation.
Nevertheless, I think it makes sense to think about the specific limitations of the current implementation regarding data classes and how they specifically affect the required functionality (ie. what would this change fix now or in the future?).
